RecoTracker/LSTCore/standalone/analysis/DNN/embed_train.ipynb

```diff
@@ -0,0 +1,2847 @@
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import torch.nn as nn\n",
+    "import torch.optim as optim\n",
+    "\n",
+    "# T5 embedding net (same as before)\n",
+    "class EmbeddingNetT5(nn.Module):\n",
+    "    def __init__(self, input_dim=30, emb_dim=6):\n",
+    "        super().__init__()\n",
+    "        self.fc1 = nn.Linear(input_dim, 32); self.relu1 = nn.ReLU()\n",
+    "        self.fc2 = nn.Linear(32, 32);         self.relu2 = nn.ReLU()\n",
+    "        self.fc3 = nn.Linear(32, emb_dim)\n",
+    "    def forward(self, x):\n",
+    "        x = self.relu1(self.fc1(x))\n",
+    "        x = self.relu2(self.fc2(x))\n",
+    "        return self.fc3(x)\n",
+    "\n",
+    "# pLS embedding net\n",
+    "class EmbeddingNetpLS(nn.Module):\n",
+    "    def __init__(self, input_dim=10, emb_dim=6):\n",
+    "        super().__init__()\n",
+    "        self.fc1 = nn.Linear(input_dim, 32); self.relu1 = nn.ReLU()\n",
+    "        self.fc2 = nn.Linear(32, 32);         self.relu2 = nn.ReLU()\n",
+    "        self.fc3 = nn.Linear(32, emb_dim)\n",
+    "    def forward(self, x):\n",
+    "        x = self.relu1(self.fc1(x))\n",
+    "        x = self.relu2(self.fc2(x))\n",
+    "        return self.fc3(x)\n",
+    "    \n",
+    "class ContrastiveLoss(nn.Module):\n",
+    "    def __init__(self, margin=1.0):\n",
+    "        super(ContrastiveLoss, self).__init__()\n",
+    "        self.margin = margin\n",
+    "    def forward(self, d, label, weight=None):\n",
+    "        # d: [B,1], label: [B,1], weight: [B,1] or None\n",
+    "        l_sim = (1 - label) * d.pow(2)\n",
+    "        l_dis = label * (self.margin - d).clamp(min=0.0).pow(2)\n",
+    "        loss = l_sim + l_dis\n",
+    "        if weight is not None:\n",
+    "            loss = loss * weight\n",
+    "        return loss.mean()\n",
+    "\n",
+    "# contrastive loss (reuse)\n",
+    "criterion = ContrastiveLoss(margin=1.0)\n",
+    "\n",
+    "# instantiate and send to GPU/CPU\n",
+    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
+    "embed_t5 = EmbeddingNetT5().to(device)\n",
+    "embed_pls = EmbeddingNetpLS().to(device)\n",
+    "\n",
+    "# joint optimizer over both nets\n",
+    "optimizer = optim.Adam(\n",
+    "    list(embed_t5.parameters()) + list(embed_pls.parameters()),\n",
+    "    lr=0.0025\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Epoch 1/200:  JointLoss=0.4559  T5=0.2255  pLS=0.2304\n",
+      "Epoch 2/200:  JointLoss=0.3017  T5=0.1444  pLS=0.1573\n",
+      "Epoch 3/200:  JointLoss=0.2703  T5=0.1329  pLS=0.1373\n",
+      "Epoch 4/200:  JointLoss=0.2568  T5=0.1285  pLS=0.1283\n",
+      "Epoch 5/200:  JointLoss=0.2502  T5=0.1262  pLS=0.1239\n",
+      "Epoch 6/200:  JointLoss=0.2466  T5=0.1248  pLS=0.1219\n",
```

Comment:
I guess these are marginally useful

Reply:
Agreed, left this one in. Removed the others.